evm/vm/forks/sharding/contracts/validator_manager.v.py

```diff
@@ -182,78 +181,79 @@ def withdraw(validator_index: num) -> bool:
+                        ),
+                        'uint256'
                 ),
-                as_num256(self.get_validators_max_index()),
-            )
+                convert(self.get_validators_max_index(), 'uint256')
+            ),
+            'int128'
         )
     ].addr
 
 
 # Attempts to process a collation header, returns True on success, reverts on failure.
 @public
 def add_header(
-        shard_id: num,
-        expected_period_number: num,
+        shard_id: int128,
+        expected_period_number: int128,
         period_start_prevhash: bytes32,
         parent_hash: bytes32,
         transaction_root: bytes32,
         collation_coinbase: address,  # TODO: cannot be named `coinbase` since it is reserved
         state_root: bytes32,
         receipt_root: bytes32,
-        collation_number: num) -> bool:  # TODO: cannot be named `number` since it is reserved
-    zero_addr = 0x0000000000000000000000000000000000000000
+        collation_number: int128) -> bool:  # TODO: cannot be named `number` since it is reserved
 
     # Check if the header is valid
     assert (shard_id >= 0) and (shard_id < self.shard_count)
     assert block.number >= self.period_length
-    assert expected_period_number == floor(decimal(block.number / self.period_length))
+    assert expected_period_number == floor(block.number / self.period_length)
     assert period_start_prevhash == blockhash(expected_period_number * self.period_length - 1)
 
     # Check if this header already exists
-    header_bytes = concat(
-        as_bytes32(shard_id),
-        as_bytes32(expected_period_number),
+    header_bytes: bytes <= 288 = concat(
+        convert(shard_id, 'bytes32'),
+        convert(expected_period_number, 'bytes32'),
         period_start_prevhash,
         parent_hash,
         transaction_root,
-        as_bytes32(collation_coinbase),
+        convert(collation_coinbase, 'bytes32'),
         state_root,
         receipt_root,
-        as_bytes32(collation_number),
+        convert(collation_number, 'bytes32'),
     )
-    entire_header_hash = sha3(header_bytes)
+    entire_header_hash: bytes32 = sha3(header_bytes)
     assert self.collation_headers[shard_id][entire_header_hash].score == 0
     # Check whether the parent exists.
     # if (parent_hash == 0), i.e., is the genesis,
     # then there is no need to check.
-    if parent_hash != as_bytes32(0):
+    if parent_hash != convert(0, 'bytes32'):
         assert self.collation_headers[shard_id][parent_hash].score > 0
     # Check if only one collation in one period perd shard
     assert self.period_head[shard_id] < expected_period_number
 
     # Check the signature with validation_code_addr
-    validator_addr = self.get_eligible_proposer(shard_id, block.number / self.period_length)
-    assert validator_addr != zero_addr
+    validator_addr: address = self.get_eligible_proposer(shard_id, floor(block.number / self.period_length))
+    assert not not validator_addr
```

Comment:
Is this a pattern of Vyper? (two `not`?)

Reply:
Yes, see https://github.com/ethereum/vyper/blob/5b70797a31ffcbdd90b1dfd98f7b38d306bca7a0/examples/voting/ballot.v.py#L30-L32

Reply:
Is the loosely equivalent to `bool(thing)` in python?

Reply:
Yes I believe so.